src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityFissionReactor.java

```diff
@@ -0,0 +1,1104 @@
+    }
+
+    private boolean isLocked() {
+        return lockingState == LockingState.LOCKED;
+    }
+
+    private TextFormatting getLockedTextColor() {
+        if (lockingState == LockingState.LOCKED)
+            return TextFormatting.GREEN;
+        if (lockingState == LockingState.INVALID_COMPONENT)
+            return TextFormatting.RED;
+        if (lockingState == LockingState.UNLOCKED)
+            return TextFormatting.DARK_AQUA;
+        if (lockingState == LockingState.SHOULD_LOCK)
+            return TextFormatting.BLACK;
+        return getWorld().getWorldTime() % 4 >= 2 ? TextFormatting.RED : TextFormatting.YELLOW;
+    }
+
+    private void tryLocking(boolean lock) {
+        if (!isStructureFormed())
+            return;
+
+        if (lock)
+            lockAndPrepareReactor();
+        else
+            unlockAll();
+    }
+
+    @Override
+    protected void addErrorText(List<ITextComponent> list) {
+        if (lockingState != LockingState.LOCKED && lockingState != LockingState.UNLOCKED) {
+            list.add(
+                    new TextComponentTranslation("gregtech.gui.fission.lock." + lockingState.toString().toLowerCase()));
+        }
+    }
+
+    @Override
+    protected void addDisplayText(List<ITextComponent> list) {
+        super.addDisplayText(list);
+        list.add(
+                TextComponentUtil.setColor(new TextComponentTranslation(
+                        "gregtech.gui.fission.lock." + lockingState.toString().toLowerCase()), getLockedTextColor()));
+        list.add(new TextComponentTranslation("gregtech.gui.fission.k_eff", String.format("%.4f", this.kEff)));
+        list.add(new TextComponentTranslation("gregtech.gui.fission.depletion",
+                String.format("%.2f", this.fuelDepletionPercent * 100)));
+    }
+
+    protected boolean isBlockEdge(@NotNull World world, @NotNull BlockPos pos, @NotNull EnumFacing direction,
+                                  int steps) {
+        BlockPos test = pos.offset(direction, steps);
+
+        if (world.getBlockState(test).getBlock() == MetaBlocks.FISSION_CASING) {
+            return false;
+        }
+
+        MetaTileEntity potentialTile = GTUtility.getMetaTileEntity(world, test);
+        if (potentialTile == null) {
+            return true;
+        }
+
+        return !(potentialTile instanceof IFissionReactorHatch || potentialTile instanceof IMaintenanceHatch);
+    }
+
+    protected EnumFacing getUp() {
+        return RelativeDirection.UP.getRelativeFacing(frontFacing, upwardsFacing, isFlipped);
+    }
+
+    protected EnumFacing getRight() {
+        return RelativeDirection.RIGHT.getRelativeFacing(frontFacing, upwardsFacing, isFlipped);
+    }
+
+    /**
+     * Uses the upper layer to determine the diameter of the structure
+     */
+    protected int findDiameter(int heightAbove) {
+        int i = 1;
+        while (i <= 15) {
+            if (this.isBlockEdge(this.getWorld(), this.getPos().offset(getUp(), heightAbove),
```

Comment:
Use a `MutableBlockPos` with `move` instead of calling `offset`. This burns a lot of memory allocations right now.